bin/circle/circle_project.rb

```diff
@@ -20,8 +23,40 @@ def initialize(project = 'code-dot-org/code-dot-org')
   # @return [Object] build descriptor object from the CircleCI API
   #   Example output JSON: https://gist.github.com/bcjordan/02f7c2906b524fa86ec75b19f9a72cd2
   def get_build(build_num, ensure_full_summary = false)
-    (!ensure_full_summary && get_recent_builds.find {|b| b['build_num'] == build_num}) ||
-        JSON.parse(open("#{@project_api_base}/#{build_num}").read)
+    # First, check local cache
+    if File.exist? cache_file_name(build_num)
+      Zlib::GzipReader.open(cache_file_name(build_num)) do |file|
+        return JSON.parse(file.read)
+      end
+    end
+
+    # Second, try pulling from the build summary
+    unless ensure_full_summary
+      short_summary = get_recent_builds.find {|b| b['build_num'] == build_num}
+      return short_summary if short_summary
+    end
+
+    # Finally, download the whole thing
+    body = nil
+    download_attempts_remaining = 3
+    while body.nil? && download_attempts_remaining > 0
+      begin
+        body = open("#{@project_api_base}/#{build_num}").read
+      rescue
+        download_attempts_remaining -= 1
+        sleep(0.25)
+      end
+    end
+
+    # Parse build and return it, saving to local cache if the build outcome is determined
+    build_summary = JSON.parse(body)
+    if build_summary && !build_summary['outcome'].nil?
+      FileUtils.mkdir_p(CACHE_DIRECTORY)
+      Zlib::GzipWriter.open(cache_file_name(build_num)) do |file|
```

Comment:
Maybe crazy util idea for future scripts (not a great fit here since we also check existence), wonder if `with_cached_data(filename) do |data|` and `write_cached_data(filename, data)` helpers could be useful in the future.